bittensor/metagraph.py

```diff
@@ -96,45 +143,65 @@ def do_gossip(self):
 
         synapses = self.synapses(1000)
         peers = self.peers(10)
-        metagraph_address = random.choice(list(self._peers))
-        realized_address = metagraph_address
-        if metagraph_address.split(':')[0] == self._config.remote_ip:
+        random_peer = random.choice(list(self._peers))
+        random_peer_address = random_peer['address']
+        realized_address = random_peer_address
+        if random_peer_address.split(':')[0] == self._config.remote_ip:
             realized_address = 'localhost:' + str(
-                metagraph_address.split(":")[1])
-        try:
-            channel = grpc.insecure_channel(realized_address)
-            stub = bittensor_grpc.MetagraphStub(channel)
-            request = bittensor_pb2.GossipBatch(peers=peers, synapses=synapses)
-            response = stub.Gossip(request, timeout=0.5)
-            self._sink(response)
-        except Exception as e:
-            # Faulty peer.
-            logger.info("Faulty peer!: {}".format(metagraph_address))
-            logger.error("ERROR: {}".format(e))
-            self._peers.remove(metagraph_address)
+                random_peer_address.split(":")[1])
+        
+        retries = 0
+        peer_reached = False
+        while (retries < 3):
+            try:
+                channel = grpc.insecure_channel(realized_address, options=(('grpc.enable_http_proxy', 0),))
+                stub = bittensor_grpc.MetagraphStub(channel)
+                request = bittensor_pb2.GossipBatch(peers=peers, synapses=synapses, source_neuron_key=self._config.neuron_key)
+                response = stub.Gossip(request)
+                channel.close()
+                self._sink(response)
+                peer_reached = True
+            except Exception as e:
+                # Faulty peer.
```

Comment:
I couldn't use the @retry decorator here because for some reason there's some incompatibility between that and GRPC, so I had to set up the retry logic manually.